### ISSUE TITLE
LFRic-specific Dependency Analysis

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#2293]
As part of #2197, we are moving to using pointers to pass data to kernel calls so as to avoid accessing structures within an OpenACC region (e.g. `my_proxy%data`). However, the Dependence Analysis now sees that we are passing an array to a kernel and, by default, we give all such arguments read-write access (`Call.reference_accesses`). It therefore correctly identifies that this is a potential race condition (since it doesn't know which elements of the array are written to) and says the call cannot be safely executed in parallel.

Since the call is actually to an LFRic kernel (which are only permitted to write to data in the current column), we can do better:

 * If the field is on a discontinuous function space then the access is safe;
 * If the parent loop is over dofs then the access is safe;
 * If the parent loop is over cells of a single colour then the access is safe;

The potential solution to this is to encapsulate this domain knowledge in a domain specialisation of the dependence analysis.

**Comment by sergisiso** [comment]
I agree that we need this additional kernel/domain information but I am not a fan of this solution:

> The potential solution to this is to encapsulate this domain knowledge in a domain specialisation of the dependence analysis.

The parallel specialization hierarchies are hard to manage. Does a generic transformation using dependency analysis would need the specialized DependencyTools when applying it to LFRic. Then we need to pass them are parameters has have it as state, or have to create specialised transforamtions (both are not ideal).

What about making the can_loop_be_parallelise be part of the loop node? Then it is the already existing LFRicLoop specialisation in the domain PSyIR that adds the logic.

**Comment by hiker** [comment]
Do you mean to just add a simple wrapper function to loop that calls the API-specific DA version? That should. work. I wouldn't want the whole code for that moved into the loop class.

I am not sure how much of a change in the DA tools would be required, i.e. if it's even worth having an API-specific version. Wouldn't it be a simple test to be added? `if isinstance(var_access.node, LFRicKern)` then the special handling?

**Comment by sergisiso** [comment]
> Do you mean to just add a simple wrapper function to loop that calls the API-specific DA version? 

No, I was hoping not to have API-specific DA versions. And have the differences just in the specialised nodes.

>  I wouldn't want the whole code for that moved into the loop class.

It doesn't need to necessarily. Just the method/functionality to distinguish between "safe"/"independent" writes and "unsafe" writes that could end up in race conditions. Maybe they should be two different kind of writes in the reference_accesses itself?

Although I wouldn't mind moving the code either, as at the moment it creates cyclic imports with Loops and always end up with inlined imports.

**Comment by arporter** [comment]
I don't think we want API-specific branches inside the DA tools. I've tried moving the `can_loop_be_parallelised` method into `Loop` and it fits quite naturally and removes the need to pass the 'loop' node into the method (which I always take to be a good sign). The only problem is the nice message-capture mechanism that the DA tools has - we lose that if we move the method out of DA tools and into Loop. We could alter the interface so that, if messages are required, an object must be passed by the caller in which to collect them?

**Comment by arporter** [comment]
In case it aids the discussion, I also note that `can_loop_be_parallelised` is only called from `ParallelLoopTrans.validate()` where we already have some LFRic specialisation:
https://github.com/stfc/PSyclone/blob/ba747755e0a35c2812e614803f7cd4bf1122cbd3/src/psyclone/psyir/transformations/parallel_loop_trans.py#L151C1-L172

**Comment by hiker** [comment]
Note that `can_loop_be_parallelised` might also be called from external scripts (e.g. my ROMS omp-parallelisation script does).

**Comment by arporter** [comment]
Ah! What's the usecase for that Joerg? (i.e. why not just call `validate()`?)

**Comment by hiker** [comment]
> I don't think we want API-specific branches inside the DA tools. I've tried moving the `can_loop_be_parallelised` method into `Loop` and it fits quite naturally and removes the need to pass the 'loop' node into the method (which I always take to be a good sign). 

Maybe - there is a lot of code underneath that function which might be useful to be reused (e.g. for fusing loop, which is not properly tested atm iirc). So I am ok with moving the high level stuff out, but I would prefer if the rest stays in the DT tools. Maybe even a slightly different approach:
what about calling a method in Loop that tests if the loop can be parallelised with regards to the specified variable. The the DT would just call this function at the start (and skip the rest of the logic). That could then easily be overwritten for each domain-specific loop.

And we could easily keep the message capability.

Otherwise passing an object through to collect messages would be ok - but all in all I would prefer to keep the PSyIR nodes small with regards to DA etc.

**Comment by hiker** [comment]
> Ah! What's the usecase for that Joerg? (i.e. why not just call `validate()`?)

I use this to determine code regions that can be combined under one `omp parallel` worksharing construct - collect all consecutive nodes, then apply `omp do` to all of them and then `omp parallel` outside.

Maybe  (probably ;) ) `validate` would work - though as far as I can see I won't be getting the messages back, which I find rather useful.

**Comment by sergisiso** [comment]
> Maybe - there is a lot of code underneath that function which might be useful to be reused (e.g. for fusing loop, which is not properly tested atm iirc). So I am ok with moving the high level stuff out, but I would prefer if the rest stays in the DT tools. 

Following this idea, I always thought that the can_[loop_]be_parallelised calling is_loop_suitable_for_parallel which does:
```
        if only_nested_loops:
            all_loops = self.walk(Loop)
            if len(all_loops) == 1:
                self._add_message("Not a nested loop.",
                                  DTCode.INFO_NOT_NESTED_LOOP)
                return False

        if self._loop_types_to_parallelise:
            if self.loop_type not in self._loop_types_to_parallelise:
                self._add_message(f"Loop has wrong loop type '"
                                  f"{self.loop_type}'.",
                                  DTCode.INFO_WRONG_LOOP_TYPE)
                return False
```
was the other way around.

I do think that the is_loop_suitable_for_parallel is domain-specific and it should call the more "fundamental" can_be_parallel, which we could even rename to are_iterations_independent or something more loop-centric.

**Comment by arporter** [comment]
One of the things I hit when beginning implementation of an LFRic-specific version of the DA was that I had to override `_is_scalar_parallelisable` to allow for scalar arguments to LFRic kernels. This is because any kernel argument is currently given 'READWRITE' access. Therefore, one solution to this particular problem is to improve LFRicKern.reference_accesses() to use the kernel metadata.

**Comment by sergisiso** [comment]
But then the loop_types and nested_loop I would argue that is a transformation specialization and not  (e.g. we may want nested loops in OpenMP but not in vectorization or gpu)

> This is because any kernel argument is currently given 'READWRITE' access. Therefore, one solution to this particular problem is to improve LFRicKern.reference_accesses() to use the kernel metadata.

I think reference_accesses needs to be improved as you suggest, the reference_accesses is used from quite a few more places and ultimately we want to use it to replace the distributed memory dependencies analysis, so we need the exact dependencies.

**Comment by hiker** [comment]
> ...
> was the other way around.
> 
> I do think that the is_loop_suitable_for_parallel is domain-specific and it should call the more "fundamental" can_be_parallel, which we could even rename to are_iterations_independent or something more loop-centric.

This function is a first filter to discard loops that the user does not want to parallelise (e.g. because there might not be enough work if the loop is not nested). It should be a cheap and quick test. Using this to call `can_be_parallelised` would potentially (depending on when it is called) mean that a lot of time is wasted in a DA that the user does not even want to.

I am happy to make this domain-specific, that makes sense (e.g. testing for nested loops in LFRic - though that's why this test can be disabled), but moving this sounds potentially like a good idea - though of course we still need a way to overwrite this.

**Comment by hiker** [comment]
The more I think about this:

> ... removes the need to pass the 'loop' node into the method (which I always take to be a good sign). 

I would not in general agree with this. Following this 'good sign' philosophie, we should move the code that converts Fortran code for a loop from the FortranWriter into the Loop, since it would need to remove the need to pass the loop node in. First of all, the loop node is 'self', so it is still there (though maybe less explicit?). Secondly, this would result in the Loop class to collect way too much code, code that is better of combined in one other functional unit.

I am strongly in favour of the Loop class providing a 'callback' that the DA can use. Maybe as one function returning if the loop must not be parallelised (dependencies would create incorrect code), should not (e.g. not nested in gocean), or should be parallelised (otherwise).

**Comment by sergisiso** [comment]
> This function is a first filter to discard loops that the user does not want to parallelise (e.g. because there might not be enough work if the loop is not nested)

That's fine but its still transformation specific I think (e.g. in GPUs you almost never want to do this since not parallelising one loop has a big penalty) as its left in the CPU or serialized/very slow on the CPU.

> Using this to call can_be_parallelised would potentially (depending on when it is called) mean that a lot of time is wasted in a DA that the user does not even want to.

Yes, that's a good point, but as you point out the key is when it is called, the transforamtions that want it could do
```
if only_nested_loops and len(all_loops) == 1:
    # Error message
    return False
if self.loop_type not in self._loop_types_to_parallelise:
    # Error message
    return False
return loop.can_be_parallel/are_iterations_independent()
```

**Comment by arporter** [comment]
Hi @sergisiso, just so that I can be sure, when you say "avoid the inlined import everywhere" what precisely do you mean?

**Comment by sergisiso** [comment]
Oh, ignore me, I thought in the past I had to bring the DP import inside the code due to its cyclic import with the loop node. But looking at the code I see this is not true :)
Still I think loop."something" is easier than importing and initialising the DP object when we want to know if iterations are independent.

**Comment by arporter** [comment]
On the slightly separate topic of `reference_accesses`, It turns out that `DynKern.reference_accesses` does work but that `DynKern` is not an instance of `Call`. When we use the generic dependence analysis on the LFRic PSy layer, we end up using `Call.reference_accesses` rather than `DynKern.reference_accesses`. I need to go back to my 'field-dereferencing' branch and check this again.

**Comment by sergisiso** [comment]
I have a branch where I convert all Kerns to inherit from Calls, but it was not passing all tests yet. Let me know if it could help.

Also I think the problem you will have with DynKern.reference_accesses is that we just say that a field is WRITE but not if its a SAFE-WRITE (only updates points that only itself can access in this kernel) or an UNSAFE-WRITE (it updates points that other iterantion access calling the same kernel). I think we need to add this distinction (and this may be a good place because we can know from the domain/metadata instead of analysing the kernel code)

**Comment by sergisiso** [comment]
BTW, the safe-write is not exclusive to kernels, we don't have atomics yet but for instance:
```
do i = 2, N
  a(i+1) = a(i) + 1
end do

do i = 2, N
  !$omp atomic
  a(i+1) = a(i) + 1
end do
```
We now just say a is WRITE and can_be_parallel is FALSE. So I think the AccessType enum should have a SAFEWRITE AccessType that the dependency analysis can use to be smarter. Kernels call in non-contiguous fields have the same AccessType for that field.
What does @hiker think?

**Comment by arporter** [comment]
> Also I think the problem you will have with DynKern.reference_accesses is that we just say that a field is WRITE but not if its a SAFE-WRITE (only updates points that only itself can access in this kernel) or an UNSAFE-WRITE (it updates points that other iterantion access calling the same kernel). I think we need to add this distinction (and this may be a good place because we can know from the domain/metadata instead of analysing the kernel code)

Yes, I was going to use domain-specific information to handle that - i.e. check that the kernel call is within a coloured loop if the write is to a field on a continuous space. However, that last bit would be much better handled by your suggested addition of a `SAFE-WRITE` access pattern.

**Comment by arporter** [comment]
I've just gone back to my field-dereferencing branch and investigated what doesn't work in more detail. Many of the test failures are because the various transformations use a generic `validate` method (e.g. of ParallelLoopTrans). Having to add a domain-specific validate for the various transformations is not particularly nice. Therefore, putting this intelligence into a lightweight method in Loop (that mostly just uses the existing DependenceAnalysis tooling) is a good way to go. Some failures will also immediately be avoided if we add the `SAFE-WRITE` access pattern into the dependence analysis, as suggested by Sergi above. Obviously, that won't fix all issues because there will be times when we will know that a WRITE is safe only when colouring is used. This intelligence will have to go in the new `DynLoop.something` method proposed by Sergi.

**Comment by hiker** [comment]
I don't like the idea of adding a `SAFE-WRITE`. IMHO, it should stay a `WRITE` access should stay exactly that, and the DA can then use API-specific information to considering this a non-overlapping write. There is a function in the DA that checks this (`_array_access_parallelisable`), and it has the variable access info object, which includes the node (which should be the kernel call).

And as far as I can see `ParallelLoopTrans` already uses the DA - so if the DA just uses a callback to the kernel call to get the loop specific information, shouldn't it all work out?

**Comment by hiker** [comment]
To make my case a bit stronger: existing DAs (like the one system I used) work with read and write accesses. Parallelisation is then determined based on if the writes overlaps or not. In these DA systems, that's based on the indices used, but in our case we can't determine this (since the indices in LFRic are using indirect addressing using the dof_map arrays), but we can make this decision based on domain-specific knowledge we have - and that's what we should be doing.

If we start adding a new type of access, we need to re-evaluate any other existing algorithm. E.g., how does SAFE-WRITE change rules for checking loop fusion? Loop swapping? Loop splitting? I would very much prefer to stay with existing, proven algorithms, and add the domain-specific knowledge in this existing framework. It's very easy to make a mistake here if we are going out on our own. After all, our goal is NOT to develop a new and better DA, all we want is to add domain-specific knowledge to an existing framework. Let the DA detect the problem, which can then be evaluated with domain-specific knowledge (not sure if we really need a two-phase handling, I think if we can handle the problem immediately when it is detected). 

In an ideal world, I would actually say we should add additional information about the dof-map arrays (which is domain-specific), and use this knowledge to determine that the write accesses are independent. We could then potentially even check that the kernels actually to not violate the API-specific conditions (which is still a scary thing - when looking at swapping loops, I noticed that there can be quite complex kernels, and it is so easy to mix up indices - to a degree that I thought a debug-transformation that verifies accesses at runtime, like a range-check, just a bit cheaper).

Re @sergisiso 's example: this is using additional knowledge, the fact that this is an atomic. Again, the DA should first detect that there is an overlap, and add the additional handling for the accesses that have a problem (which as far as I can see are not domain-specific - this is an odd case, you would use the DA to determine that there is a conflict, and the DA should tell you that there is a write-conflict of writing to `a(i+1)` to itself. Then an atomic is added, and if this is the only problem reported by the DA, we can parallelise the loop (and disable the DA, since we tested it previously).
